### PR TITLE
Clean up flake8 warnings from doctests

### LINKED ISSUE #19374: Clean up flake8 warnings from doctests
Currently flake8 runs on the main code but not on doctests. It has a `--doctests` argument to check the doctests as well.

For example:
```console
$ flake8 --doctests sympy/matrices/determinant.py 
sympy/matrices/determinant.py:545:9: F401 'sympy.MatrixSymbol' imported but unused
sympy/matrices/determinant.py:545:9: F401 'sympy.eye' imported but unused
sympy/matrices/determinant.py:545:9: F401 'sympy.det' imported but unused
```
That can be fixed with
```diff
diff --git a/sympy/matrices/determinant.py b/sympy/matrices/determinant.py
index 647541afef..f3bb9a4067 100644
--- a/sympy/matrices/determinant.py
+++ b/sympy/matrices/determinant.py
@@ -542,7 +542,7 @@ def _det(M, method="bareiss", iszerofunc=None):
     Examples
     ========
 
-    >>> from sympy import Matrix, MatrixSymbol, eye, det
+    >>> from sympy import Matrix
     >>> M = Matrix([[1, 2], [3, 4]])
     >>> M.det()
     -2
```

Running flake8 on the whole of sympy with
```
$ flake8 --doctests sympy
```
leads to 698 warnings almost all of which are F401 (unnecessary import) and so easily fixable.

These should all be fixed and then we should add the `--doctest` argument on Travis.

I'm marking this as easy to fix.

### PR DESCRIPTION
**Pull request by InCogNiTo124**
#### References to other Issues or PRs

Fixes #19374 

#### Brief description of what is fixed or changed
running `flake8 --doctests sympy/` returns a bunch of errors. This PR is to fix them

#### Release Notes

* other
  * Improve code quality by cleaning the doctests up

### CODE REVIEW

**Comment by InCogNiTo124** [comment]
Most of the unused imports were deleted, but some docstrings were improved, such as `matrices/common.py`, `functions/elementary/trigonometric.py` and some others

also, I only got rid od all F401, I've yet to get rid of other errors (F403, F811, F821)

p.s. sorry for the deleted file, I did it accidentally somehow in terminal but I restored it back immediately

**Comment by oscarbenjamin** [comment]
One of the commits deletes the travis fail and another adds it back. If this is merged then that file would be deleted for one commit in the history and would then lose all of its history as view through `git blame`. The commits here need to be squashed so that there is no commit that removes that file.

**Comment by InCogNiTo124** [comment]
I think all errors that are left are those that populate the namespace directly, and frankly I'm not really sure how to approach in handling those other than `#noqa`.

**Comment by oscarbenjamin** [comment]
There is a merge conflict that needs to be fixed.

By the way you should not use your master branch for writing a pull request.

**Comment by InCogNiTo124** [comment]
> By the way you should not use your master branch for writing a pull request.

Thanks for the tip, I'm still learning. Will this be a serious problem?

**Comment by oscarbenjamin** [comment]
> > By the way you should not use your master branch for writing a pull request.
> 
> Thanks for the tip, I'm still learning. Will this be a serious problem?

The most obvious problem is it makes it difficult to write two PRs simultaneously. Your master branch is now tied to this PR until it gets merged (some PRs never get merged!).

Afterwards though it means that the history on your master branch no longer corresponds to the history of the main repo. There are ways to fix this but I suggest leaving it for now until this gets merged (I think it will).

Step 1 is to fix the merge conflicts with master (i.e. the main repo master - if you didn't commit to your master then there would be no distinction).

**Comment by InCogNiTo124** [comment]
I can't put `#noqa` in `interactive/session.py` nor `utilities/_compilation/compilation.py` because the tests that fail overthere are annotated with `#doctest: +SKIP` so `doctest` module doing the actual testing interprets it as its option (which then makes it fail when running `./bin/doctest`).

**Comment by oscarbenjamin** [comment]
The doctest for `compile_link_import_py_ext` is broken so flake8 is right to report an error there. I suggest just to remove the doctest there.

For session.py I suggest ignoring those errors for the file:
https://stackoverflow.com/a/54454433/9450991

**Comment by InCogNiTo124** [comment]
Something subtle is happening in `core/basic.py:579`, this is the only doctest to be failing

**Comment by oscarbenjamin** [comment]
That doctest should give False but I suspect that the surrounding text would have been changed on master as well

**Comment by oscarbenjamin** [comment]
I think that in resolving the merge conflict you've re-added part of the docstring that was removed on master:
```diff
diff --git a/sympy/core/basic.py b/sympy/core/basic.py
index 53beba2396..14b5fceb0d 100644
--- a/sympy/core/basic.py
+++ b/sympy/core/basic.py
@@ -554,7 +554,7 @@ def as_dummy(self):
         ========
 
         >>> from sympy import Integral, Symbol
-        >>> from sympy.abc import x, y
+        >>> from sympy.abc import x
         >>> r = Symbol('r', real=True)
         >>> Integral(r, (r, x)).as_dummy()
         Integral(_0, (_0, x))
@@ -570,6 +570,14 @@ def as_dummy(self):
         a property, `bound_symbols` that returns those symbols
         appearing in the object.
 
+        Lambda and Subs have bound symbols, but because of how they
+        are cached, they already compare the same regardless of their
+        bound symbols:
+
+        >>> from sympy.abc import x, y
+        >>> from sympy import Lambda
+        >>> Lambda(x, x + 1) == Lambda(y, y + 1)
+        True
         """
         from sympy.core.symbol import Dummy, Symbol
         def can(x):
```

**Comment by oscarbenjamin** [comment]
This is how the docstring looks on current master:
https://github.com/sympy/sympy/blob/b0384246d34381a189922b19b2a82a447fc9d180/sympy/core/basic.py#L546-L573

**Comment by InCogNiTo124** [comment]
Is it done now?

**Comment by oscarbenjamin** [comment]
This has merge conflicts again already. This is the difficulty with a PR that touches so many parts of the codebase.

Also one of the earliest commits deletes a file which is then readded later. Unfortunately it is not enough to add the file in a subsequent commit as this removes all of the "history" from the file. 

Compare master:
https://github.com/sympy/sympy/blame/master/bin/test_travis.sh
With the PR:
https://github.com/sympy/sympy/blame/e6cc7fa5902c8e3673f9a5b1f45332862489c3ae/bin/test_travis.sh

The commits will need to be squashed or recommitted so that the file is not removed in the history.

File: bin/test_travis.sh [hunk 1]
```diff
@@ -7,6 +7,7 @@ set -x
 
 if [[ "${TEST_FLAKE8}" == "true" ]]; then
     flake8 sympy;
+    flake8 --doctests --per-file-ignores='sympy/interactive/session.py:F821' sympy;
```
Comment (oscarbenjamin) [inline bin/test_travis.sh:10]: Does this run the normal flake8 checking twice?

**Comment by oscarbenjamin** [comment]
There were problems with the way that the commits had been squashed in combination with merging with master. I think I've resolved them and force-pushed onto this PR.

You should use these commands to synchronise the changes to your computer:
```
git fetch origin
git reset origin/master --hard
```
(Don't use `git push` as it will try to merge)

File: bin/test_travis.sh (hunk 1 above)
Comment (InCogNiTo124) [reply bin/test_travis.sh:10]: Yes, I found no way for flake to simoultaneously check for docs and code

Comment (InCogNiTo124) [reply bin/test_travis.sh:10]: I'm not sure. It seems so, but I can't find anything in the `flake8` documentation that says whether or not the code is also checked with the `--doctests` flag.

edit: I found a reference here: https://bugs.launchpad.net/pyflakes/+bug/1223150 so it seems that the normal checking truly is done twice.

**Comment by oscarbenjamin** [comment]
Can the per-file-ignores be put in setup.cfg?

**Comment by InCogNiTo124** [comment]
> Can the per-file-ignores be put in setup.cfg?

[Yes](https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-per-file-ignores). Should I refactor that option there?

**Comment by oscarbenjamin** [comment]
Yes. There already others there

**Comment by oscarbenjamin** [comment]
I've just been through the whole diff. This is good to merge once tests pass (it should be merged quickly before new conflicts arise)

**Comment by oscarbenjamin** [comment]
Thanks @InCogNiTo124! I know it's not easy to run through the whole codebase like this.

This is merged now.

**Comment by InCogNiTo124** [comment]
> Thanks @InCogNiTo124! I know it's not easy to run through the whole codebase like this.
> 
> This is merged now.

This was my first opensource PR and it as truly a baptism by fire, but I really learned a lot :D

**Comment by oscarbenjamin** [comment]
You've handled this very well. I look forward to seeing more :)

Now that this is in you should fix your master branch e.g. with:
```
git reset HEAD~100 --hard
git fetch origin
git pull
```
For future PRs create a branch for each PR so you can keep your master branch clean.

**Comment by InCogNiTo124** [comment]
Thanks :) I'll come back for some more easy fixes

**Comment by oscarbenjamin** [comment]
The instruction I just gave won't work actually because your master is pushed to origin (your fork). You need to add the main sympy repo as a remote (called upstream) and then do:
```
git reset HEAD~100 --hard
git fetch upstream
git merge upstream/master
```

**Pull request merged**